src/stages/headers.rs

```diff
@@ -631,37 +636,26 @@ impl HeaderDownload {
     }
 
     #[inline]
-    fn check_headers(node: &Node, headers: Vec<BlockHeader>) -> Vec<(H256, BlockHeader)> {
-        let mut headers = headers
+    fn check_headers(
+        consensus: &Arc<dyn Consensus>,
+        headers: Vec<BlockHeader>,
+    ) -> Result<Vec<(H256, BlockHeader)>, DuoError> {
+        let headers = headers
             .into_iter()
             .map(|h| (h.hash(), h))
             .collect::<Vec<_>>();
-
-        if let Some(valid_till) =
-            headers
-                .iter()
-                .skip(1)
-                .enumerate()
-                .position(|(i, (hash, header))| {
-                    header.parent_hash != headers[i].0
-                        || header.number != headers[i].1.number + 1u8
-                        || node.bad_blocks.contains(hash)
-                })
-        {
-            headers.truncate(valid_till);
+        for (i, _) in headers.iter().enumerate().skip(1) {
+            consensus.validate_block_header(&headers[i].1, &headers[i - 1].1, false)?;
         }
-        headers
+
+        Ok(headers)
     }
 
     fn verify_seal(&self, headers: &mut Vec<(H256, BlockHeader)>) {
```

Comment:
predicate invocation of all of this on `Consensus::needs_parallel_validation` (see the relevant comment)